examples/step-74/step-laplace-FEIV.cc

```diff
@@ -0,0 +1,1245 @@
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/fe/mapping_q1.h>
+// Here the discontinuous finite elements are defined. They are used in the
+// same way as all other finite elements, though -- as you have seen in
+// previous tutorial programs -- there isn't much user interaction with finite
+// element classes at all: they are passed to <code>DoFHandler</code> and
+// <code>FEValues</code> objects, and that is about it.
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_interface_values.h>
+// We are going to use the simplest possible solver, called Richardson
+// iteration, that represents a simple defect correction. This, in combination
+// with a block SSOR preconditioner (defined in precondition_block.h), that
+// uses the special block matrix structure of system matrices arising from DG
+// discretizations.
+#include <deal.II/lac/solver_gmres.h>
+#include <deal.II/lac/precondition_block.h>
+#include <deal.II/lac/precondition.h>
+// We are going to use gradients as refinement indicator.
+#include <deal.II/numerics/derivative_approximation.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/base/convergence_table.h>
+#include <deal.II/meshworker/mesh_loop.h>
+
+// Like in all programs, we finish this section by including the needed C++
+// headers and declaring we want to use objects in the dealii namespace
+// without prefix.
+#include <iostream>
+#include <fstream>
+
+#include "loop_stuff.h"
+
+namespace Step12
+{
+  using namespace dealii;
+
+  enum Test_Case
+  {
+    convergence_rate,
+    l_singularity
+  };
+
+
+  template <int dim>
+  class Solution:  public Function<dim>
+  {
+  public:
+    Solution(Test_Case in_tc)
+      :
+      Function<dim>(),
+      test_case(in_tc)
+      {}
+    virtual void value_list (const std::vector<Point<dim> > &points,
+                             std::vector<double> &values,
+                             const unsigned int component=0) const
+    {
+      switch (test_case)
+        {
+          case convergence_rate :
+            {
+              using numbers::PI;
+              for (unsigned int i=0; i<values.size(); ++i)
+                values[i]= std::sin(2*PI*points[i][0]) * std::sin(2*PI*points[i][1]);
+              break;
+            }
+          case l_singularity :
+            {
+              for (unsigned int i=0; i<values.size(); ++i)
+                values[i]= ref.value(points[i]);
+              break;              
+            }
+          default:
+            {
+              Assert(false, ExcNotImplemented());
+            }             
+        }
+        
+    }
+    private:
+    Functions::LSingularityFunction ref;
+    Test_Case test_case;
+
+  };
+
+
+
+  template <int dim>
+  class Viscosity:  public Function<dim>
```

Comment:
if viscosity is 1, we might not need it as a class. Thoughts?

Reply:
Having a non-unit viscosity is only pertinent if you have an advection-diffusion term, in this case viscosity only changes the scaling of f when you think about it.